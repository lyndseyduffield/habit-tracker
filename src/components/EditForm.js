import moment from "moment";
import React, { useState } from "react";
import useForm from "react-hook-form";
import DatePicker from "react-datepicker";
import { editHabit } from "../actions";
import { connect } from "react-redux";
import "react-datepicker/dist/react-datepicker.css";
import { updateHabitStreak } from "../utils";

const EditForm = props => {
  const [state, setState] = useState({
    startDate: new Date(),
    endDate: new Date()
  });
  const { register, handleSubmit, setValue, errors } = useForm();

  React.useEffect(() => {
    register({ name: "startDate" });
    register({ name: "endDate" });
  }, [register, setValue]);

  React.useEffect(() => {
    if (props.habit) {
      const startDate = props.habit.startDate.toDate();
      const endDate = props.habit.endDate.toDate();
      setValue("startDate", startDate);
      setValue("endDate", endDate);
      setState({ startDate, endDate });
    }
  }, [setValue, props.habit]);

  const handleDatePickerChange = (key, date) => {
    if (key === "startDate" && date > state.endDate) {
      setValue("startDate", date);
      setValue("endDate", date);
      setState({ startDate: date, endDate: date });
    } else {
      setValue(key, date);
      setState({ ...state, [key]: date });
    }
  };

  const onSubmit = data => {
    const id = props.match.params.id;
    const startDate = moment(data.startDate);
    const endDate = moment(data.endDate);
    const habit = updateHabitStreak({
      ...props.habit,
      ...data,
      startDate,
      endDate
    });
    props.dispatch(editHabit(habit, id));
    props.history.push("/");
  };

  return (
    <form class="container" onSubmit={handleSubmit(onSubmit)}>
      <div class="field">
        <label class="label">Habit</label>
        <div class="control">
          <input
            class="input"
            type="text"
            name="title"
            defaultValue={props.habit && props.habit.title}
            ref={register({ required: true })}
          />
        </div>
        <p class="help is-danger">{errors.title && "A title is required"}</p>
      </div>
      <div class="field">
        <label class="label">Goal</label>
        <div class="control">
          <input
            class="input"
            type="text"
            name="goal"
            defaultValue={props.habit && props.habit.goal}
            ref={register}
          />
        </div>
      </div>
      <div class="field is-grouped">
        <label class="label">Start Date</label>
        <div class="control">
          <DatePicker
            minDate={new Date()}
            selected={state.startDate}
            onChange={date => {
              handleDatePickerChange("startDate", date);
            }}
          />
        </div>
        <label class="label">End Date</label>
        <div class="control">
          <DatePicker
            minDate={state.startDate}
            selected={state.endDate}
            onChange={date => {
              handleDatePickerChange("endDate", date);
            }}
          />
        </div>
      </div>

      <h3 class="label is-large">Accountability Partner</h3>
      <div class="field is-grouped">
        <label class="label">Name</label>
        <div class="control">
          <input
            class="input"
            type="text"
            name="accountabilityPartner.name"
            defaultValue={props.habit && props.habit.accountabilityPartner.name}
            ref={register({
              maxLength: { value: 40, message: "This name is too long" }
            })}
          />
        </div>
        <p class="help is-danger">
          {errors["accountabilityPartner.name"] ? (
            <span>{errors["accountabilityPartner.name"].message}</span>
          ) : (
            ""
          )}{" "}
        </p>
        <label class="label">Email</label>
        <div class="control">
          <input
            class="input"
            type="text'"
            name="accountabilityPartner.email"
            defaultValue={
              props.habit && props.habit.accountabilityPartner.email
            }
            ref={register({
              pattern: {
                value: /^[^\s@]+@[^\s@]+\.[^\s@]+$/,
                message: "Please enter a valid email"
              }
            })}
          />
        </div>
        <p class="help is-danger">
          {errors["accountabilityPartner.email"] ? (
            <span>{errors["accountabilityPartner.email"].message}</span>
          ) : (
            ""
          )}
        </p>
      </div>
      <div class="control">
        <button class="button is-link">Submit</button>
      </div>
    </form>
  );
};

const mapStateToProps = (state, ownProps) => {
  return {
    habit: state.habits[ownProps.match.params.id]
  };
};

export default connect(mapStateToProps)(EditForm);
