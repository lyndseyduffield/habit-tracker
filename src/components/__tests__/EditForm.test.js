import React from "react";
import EditForm from "../EditForm";
import renderer from "react-test-renderer";
import { testReducer } from "../../reducers/__tests__/index.test";
import { createStore } from "redux";
import { Provider } from "react-redux";

const store = createStore(testReducer);

it("renders the edit form correctly", () => {
  const tree = renderer
    .create(
      <Provider store={store}>
        <EditForm
          match={{
            path: "/:id/edit",
            url: "/4/edit",
            isExact: true,
            params: { id: "4" }
          }}
        />
      </Provider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
