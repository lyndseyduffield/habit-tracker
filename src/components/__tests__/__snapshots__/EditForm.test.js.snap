// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the edit form correctly 1`] = `
<form
  class="form-container"
  onSubmit={[Function]}
>
  <div
    class="field"
  >
    <label
      class="label is-large"
    >
      Habit
    </label>
    <div
      class="control"
    >
      <input
        class="input"
        defaultValue="ertwet"
        name="title"
        type="text"
      />
    </div>
  </div>
  <div
    class="field"
  >
    <label
      class="label"
    >
      Goal
    </label>
    <div
      class="control"
    >
      <textarea
        class="textarea"
        defaultValue=""
        name="goal"
        type="text"
      />
    </div>
  </div>
  <div
    class="field is-grouped is-horizontal"
  >
    <div
      class="field-body"
    >
      <div
        class="field"
      >
        <label
          class="label"
        >
          Start Date
        </label>
        <div
          class="control"
        >
          <div
            className="react-datepicker-wrapper"
          >
            <div
              className="react-datepicker__input-container"
            >
              <input
                className=""
                disabled={false}
                onBlur={[Function]}
                onChange={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                readOnly={false}
                type="text"
                value="02/04/2020"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="field"
      >
        <label
          class="label"
        >
          End Date
        </label>
        <div
          class="control"
        >
          <div
            className="react-datepicker-wrapper"
          >
            <div
              className="react-datepicker__input-container"
            >
              <input
                className=""
                disabled={false}
                onBlur={[Function]}
                onChange={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                readOnly={false}
                type="text"
                value="02/04/2020"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <h3
    class="label is-large margin-top"
  >
    Accountability Partner
  </h3>
  <div
    class="field"
  >
    <label
      class="label"
    >
      Name
    </label>
    <div
      class="control"
    >
      <input
        class="input"
        defaultValue=""
        name="accountabilityPartner.name"
        type="text"
      />
    </div>
    <p
      class="help is-danger"
    >
      
    </p>
  </div>
  <div
    class="field"
  >
    <label
      class="label"
    >
      Email
    </label>
    <div
      class="control"
    >
      <input
        class="input"
        defaultValue=""
        name="accountabilityPartner.email"
        type="text"
      />
    </div>
    <p
      class="help is-danger"
    >
      
    </p>
  </div>
  <div
    class="control margin-top"
  >
    <button
      class="button is-link"
    >
      Submit
    </button>
  </div>
</form>
`;
